src/data_structures/fenwick.md

```diff
@@ -29,138 +29,110 @@ It was first described in a paper titled "A new data structure for cumulative fr
-\end{align}$$
-
-There exists a simple implementation using bitwise operations for the non-trivial operation described above:
-
-$$g(i) = i ~\&~ (i+1),$$
-
-where $\&$ is the bitwise AND operator. It is not hard to convince yourself that this solution does the same thing as the operation described above.
-
-Now, we just need to find a way to iterate over all $j$'s, such that $g(j) \le i \le j$.
-
-It is easy to see that we can find all such $j$'s by starting with $i$ and flipping the last unset bit.
-We will call this operation $h(j)$.
-For example, for $i = 10$ we have:
-
-$$\begin{align}
-10 &= 0001010_2 \\\\
-h(10) = 11 &= 0001011_2 \\\\
-h(11) = 15 &= 0001111_2 \\\\
-h(15) = 31 &= 0011111_2 \\\\
-h(31) = 63 &= 0111111_2 \\\\
-\vdots &
+g(7) = g(111_2) = 110_2 &= 6 \\\\
+g(6) = g(110_2) = 100_2 &= 4 \\\\
+g(4) = g(100_2) = 000_2 &= 0 \\\\
 \end{align}$$
 
-Unsurprisingly, there also exists a simple way to perform $h$ using bitwise operations:
+The last set bit can be extracted using $i ~\&~ (-i)$, so the operation can be expressed as:
 
-$$h(j) = j ~|~ (j+1),$$
+$$g(i) = i - (i ~\&~ (-i)).$$
 
-where $|$ is the bitwise OR operator.
+And it's not hard to see, that you need to change all values $T[j]$ in the sequence $i,~ h(i),~ h(h(i)),~ \dots$ when you want to update $A[j]$, where $h(i)$ is defined as:
 
-The following image shows a possible interpretation of the Fenwick tree as tree.
-The nodes of the tree show the ranges they cover.
+$$h(i) = i + (i ~\&~ (-i)).$$
 
-<center>![Binary Indexed Tree](binary_indexed_tree.png)</center>
+As you can see, the main benefit of this approach is that the binary operations complement each other very nicely.
 
 ## Implementation
 
 ### Finding sum in one-dimensional array
 
 Here we present an implementation of the Fenwick tree for sum queries and single updates.
 
-The normal Fenwick tree can only answer sum queries of the type $[0, r]$ using `sum(int r)`, however we can also answer other queries of the type $[l, r]$ by computing two sums $[0, r]$ and $[0, l-1]$ and subtract them.
+The normal Fenwick tree can only answer sum queries of the type $[1, r]$ using `sum(int r)`, however we can also answer other queries of the type $[l, r]$ by computing two sums $[1, r]$ and $[1, l-1]$ and subtract them.
 This is handled in the `sum(int l, int r)` method.
 
 Also this implementation supports two constructors.
 You can create a Fenwick tree initialized with zeros, or you can convert an existing array into the Fenwick form.
 
 
-```{.cpp file=fenwick_sum}
+```{.cpp file=fenwick_sum_onebased}
 struct FenwickTree {
     vector<int> bit;  // binary indexed tree
     int n;
 
     FenwickTree(int n) {
         this->n = n;
-        bit.assign(n, 0);
+        bit.assign(n+1, 0);
     }
 
     FenwickTree(vector<int> const &a) : FenwickTree(a.size()) {
-        for (size_t i = 0; i < a.size(); i++)
-            add(i, a[i]);
+        for (size_t i = 1; i <= a.size(); i++)
+            add(i-1, a[i-1]);
```

Comment:
This is the same as the 0-based constructor, because the API for `add` is 0-based.